man/hostname.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.13.
.TH HOSTNAME "1" "March 2014" "GNU inetutils 1.9.2-dirty" "User Commands"
.SH NAME
hostname \- show or set system host name
.SH SYNOPSIS
.B hostname
[\fIOPTION\fR...] [\fINAME\fR]
.SH DESCRIPTION
Show or set the system's host name.
.TP
\fB\-a\fR, \fB\-\-aliases\fR
alias names
.TP
\fB\-d\fR, \fB\-\-domain\fR
DNS domain name
.TP
\fB\-f\fR, \fB\-\-fqdn\fR, \fB\-\-long\fR
DNS host name or FQDN
.TP
\fB\-F\fR, \fB\-\-file\fR=\fIFILE\fR
set host name or NIS domain name from FILE
.TP
\fB\-i\fR, \fB\-\-ip\-addresses\fR
addresses for the host name
.TP
\fB\-s\fR, \fB\-\-short\fR
short host name
.TP
\fB\-y\fR, \fB\-\-yp\fR, \fB\-\-nis\fR
NIS/YP domain name
.TP
\-?, \fB\-\-help\fR
give this help list
.TP
\fB\-\-usage\fR
give a short usage message
.TP
\fB\-V\fR, \fB\-\-version\fR
print program version
.PP
Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.
.SH AUTHOR
Written by Debarshi Ray.
.SH "REPORTING BUGS"
Report bugs to <bug\-inetutils@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B hostname
is maintained as a Texinfo manual.  If the
.B info
and
.B hostname
programs are properly installed at your site, the command
.IP
.B info hostname
.PP
should give you access to the complete manual.


NEWS
GNU inetutils NEWS -- history of user-visible changes.
Copyright (C) 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005,
2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014 Free Software
Foundation, Inc.  See the end of this file for license conditions.

Please send inetutils bug reports to <bug-inetutils@gnu.org>.

December 23, 2013
Version 1.9.2:

An advisory file CHECKLIST containing testing and verification advice
has been added, pointing administrators to important properties and
settings for built executables.

* IDN -- International Domain Names

The clients `ftp', `ping', `ping6', `talk', `telnet', `tftp', and
`traceroute' are now capable of using libidn when resolving domain
names with character sets richer than ASCII.

Systems using the GNU C library achieve this ability by other means in
the present code, so they do not need to link against libidn at all.

* New tool 'dnsdomainname'.

* ftp

Called as `pftp', passive mode is preset.  New switch `-e/--no-edit'
for compatibility with other implementations.  Can now be built
without readline by passing `--disable-readline' to configure.  New
command `lpwd' avoids a shell call for a common task.  The legacy
limit for lengths of input lines (previously 200) has been removed.
Now it is system dependent and is common with most other tools.

* ifconfig

Provides new support for BSD and improved support for Solaris.
Argument handling has been corrected to avoid some segfaults.

* ping

New switches `--ip-timestamp', `--mask', `-T/--tos' and `--ttl'.  Can
use file capabilities instead of suid-setting.

* ping6

New switches `--hoplimit/--ttl', `-T/--tos' and `-v/--verbose'.  File
capabilities are sufficient.

* rcp

Supports IPv6 and Shishi.  New switches `-4/--ipv4', `-6/--ipv6',
`-k/--realm', `-K/--kerberos', and `-x/--encrypt'.  A file capability
like `cap_net_bind_service' can be applied instead of installing the
binary setuid.

* rexec

Supports IPv6.  New switches `-4/--ipv4', `-6/--ipv6', `-a/--ipany'

* rlogin, rsh

Support IPv6.  New switches `-4/--ipv4', `-6/--ipv6'.  A file
capability like `cap_net_bind_service' can be applied instead of
installing the binary setuid.

* syslogd

Messages can be logged to specific files according to the name of the
submitting program, on the tag attached to the message.  The mechanism
is patterned after same feature used in on BSD.

Support loading of multiple configuration files located in a single
directory.  New switches `-D/--rcdir'.

* telnet

New switch `-k/--realm'.

* traceroute

Default mode is now UDP, for better conformity with Unices.  New
switches `-f/--first-hop', `-g/--gateways', `-I/--icmp',
`-m/--max-hop', `-t/--tos', and `-w/--wait'.  File capabilities
suffice.

* ftpd

Full IPv6 support.  PAM support for Linux-PAM, OpenPAM, and Sun-PAM as
service `ftp'.  Observe that the latter two implementations only
affect non-anonymous access, since `pam_ftp.so' exists for Linux-PAM
only, so the standard checks are enforced on `ftp/anonymous' as usual.
The server now accepts the commands LPRT and LPSV.

The parsing of `/etc/ftpusers' and `/etc/ftpchroot' now allows the
specification of group names, simply by preceding the identifier with
an at-sign `@'.  An isolated `@' acts as a wildcard and matches every
user name.  In addition, the parser is more forgiving of spaces and
tabs, thus achieving better portability of these two files when
compared with other systems.

The server will by default not disclose that the requested user name
is invalid, but will instead proceed with a faked password dialogue.
This is in accordance with RFC 2577.  The new option `--non-rfc2577'
will now restore the legacy type responses, informing about
non-existing users and also about expired accounts, or expired
passwords.

* inetd

Service configuration now allows setting also the group.

* rexecd

Can be built with PAM support as service `rexec'.  New option
`-l/--logging'.

* rlogind

Support for IPv6 and PAM as services `rlogin' and `krlogin'.
Buildable with libwrap support for use in stand-alone daemon mode.
New switches `-4/--ipv4', `-6/--ipv6', and `-S/--server-principal'.

* rshd

Now has support for IPv6, PAM, and Shishi, with PAM services `rsh' and
`krsh'.  New switches `-k/--kerberos', `-r/--reverse-required',
`-S/--server-principal', `-v/--vacuous', and `-x/--encrypt'.

* talkd

Has seen reworked ACL-mechanisms, believed to be functional!  New
switches `-l/--logging', and `-S/--strict-policy'.

* telnetd

New switch `-S/--server-principal'.

* tftpd

New switches `-g/--group', `-s/--secure-dir', and `-u/--user'.

* uucpd

New switch `-u/--uucico'.

January 6, 2012
Version 1.9.1:

* Don't install git-merge-changelog.

December 31, 2011
Version 1.9:

* Security fixes

This release fixes a security flaw in telnetd that could lead to
remote access.

* Solaris is now a supported platform

Support is still new, so please report any bugs to
<bug-inetutils@gnu.org>.

* Many OpenBSD fixes.

* ftp, tftp, tftpd

Handle IPv6 connections.  ftpd is still IPv4-only until next release.

* syslogd, logger

Handle IPv6 connections.  The server can forward messages without
using a listening INET socket.  It can also bind to a single
INET/INET6 address.

* rlogind, rshd

This legacy code was vulnerable to an attack based on source routing.
Whenever either of IPOPT_SSRR or IPOPT_LSRR is encountered, the newly
initiated connection is now server-side aborted.  The advisory issued
by Secure Networks, Inc., February 1997, was adapted.  It was found at
http://www.citi.umich.edu/u/provos/papers/secnet-spoof.txt

* ifconfig

Now under GNU/Linux "ifconfig -a" shows also interfaces without an
address.

Adjusted output formating for "unix".

* ping

Now supports -W parameter to specify number of seconds to wait for
response.

* Various bugs fixes, internal improvements and clean ups.

May 15, 2010
Version 1.8:

* ftp

New options --active and --passive to enable/disable active transfer

src/hostname.c
/*
  Copyright (C) 2008, 2009, 2010, 2011, 2012, 2013, 2014 Free Software
  Foundation, Inc.

  This file is part of GNU Inetutils.

  GNU Inetutils is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or (at
  your option) any later version.

  GNU Inetutils is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see `http://www.gnu.org/licenses/'. */

#include <config.h>

#include <argp.h>
#include <error.h>
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <netinet/in.h>
#include <arpa/inet.h>
#include <progname.h>
#include "libinetutils.h"
#include "xalloc.h"
#include "xgethostname.h"
#include "xgetdomainname.h"


typedef struct
{
  const char *hostname_file;
  char *hostname_new;
  short int hostname_alias;
  short int hostname_fqdn;
  short int hostname_ip_address;
  short int hostname_dns_domain;
  short int hostname_short;
} hostname_arguments;

static char *(*get_name_action) (void) = NULL;
static int (*set_name_action) (const char *name, size_t size) = NULL;

const char args_doc[] = "[NAME]";
const char doc[] = "Show or set the system's host name.";
const char *program_authors[] = {
	"Debarshi Ray",
	NULL
};
static struct argp_option argp_options[] = {
#define GRP 0
  {"aliases", 'a', NULL, 0, "alias names", GRP+1},
  {"domain", 'd', NULL, 0, "DNS domain name", GRP+1},
  {"file", 'F', "FILE", 0, "set host name or NIS domain name from FILE",
   GRP+1},
  {"fqdn", 'f', NULL, 0, "DNS host name or FQDN", GRP+1},
  {"long", 'f', NULL, OPTION_ALIAS, "DNS host name or FQDN", GRP+1},
  {"ip-addresses", 'i', NULL, 0, "addresses for the host name", GRP+1},
  {"short", 's', NULL, 0, "short host name", GRP+1},
  {"yp", 'y', NULL, 0, "NIS/YP domain name", GRP+1},
  {"nis", 0, NULL, OPTION_ALIAS, NULL, GRP+1},
#undef GRP
  {NULL, 0, NULL, 0, NULL, 0}
};

static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  hostname_arguments *const args = (hostname_arguments *const) state->input;

  switch (key)
    {
    case 'a':
      get_name_action = xgethostname;
      args->hostname_alias = 1;
      break;

    case 'd':
      get_name_action = xgethostname;
      args->hostname_fqdn = 1;
      args->hostname_dns_domain = 1;
      break;

    case 'F':
      set_name_action = sethostname;
      args->hostname_file = arg;
      break;

    case 'f':
      get_name_action = xgethostname;
      args->hostname_fqdn = 1;
      break;

    case 'i':
      get_name_action = xgethostname;
      args->hostname_ip_address = 1;
      break;

    case 's':
      get_name_action = xgethostname;
      args->hostname_fqdn = 1;
      args->hostname_short = 1;
      break;

    case 'y':
      get_name_action = xgetdomainname;
      break;

    case ARGP_KEY_ARG:
      set_name_action = sethostname;
      args->hostname_new = strdup (arg);
      if (args->hostname_new == NULL)
        error (EXIT_FAILURE, errno, "strdup");
      break;

    default:
      return ARGP_ERR_UNKNOWN;
    }

  return 0;
}

static struct argp argp =
  {argp_options, parse_opt, args_doc, doc, NULL, NULL, NULL};

static void get_name (const hostname_arguments *const args);
static void set_name (const hostname_arguments *const args);
static char * get_aliases (const char *const host_name);
static char * get_fqdn (const char *const host_name);
static char * get_ip_addresses (const char *const host_name);
static char * get_dns_domain_name (const char *const host_name);
static char * get_short_hostname (const char *const host_name);
static char * parse_file (const char *const file_name);

int
main (int argc, char *argv[])
{
  hostname_arguments args;

  set_program_name (argv[0]);

  memset ((void *) &args, 0, sizeof (args));

  /* Parse command line */
  iu_argp_init ("hostname", program_authors);
  argp_parse (&argp, argc, argv, 0, NULL, (void *) &args);

  /* Set default action */
  if (get_name_action == NULL && set_name_action ==  NULL)
    get_name_action = xgethostname;

  if (get_name_action == xgetdomainname || get_name_action == xgethostname)
    get_name (&args);
  else if (set_name_action == sethostname)
    set_name (&args);

  return 0;
}

static void
get_name (const hostname_arguments *const args)
{
  char *sname, *name;

  sname = (*get_name_action) ();

  if (!sname)
    error (EXIT_FAILURE, errno, "cannot determine name");

  if (args->hostname_alias == 1)
    name = get_aliases (sname);
  else if (args->hostname_fqdn == 1)
    {
      name = get_fqdn (sname);

      if (args->hostname_dns_domain == 1 || args->hostname_short == 1)
	{
	  /* Eliminate empty replies, as well as `(none)'.  */
	  if (name && *name && *name != '(')
	    {
	      free (sname);
	      sname = name;
	      name = NULL;
	    }
	  else if (name && *name == '(')
	    {
	      free (name);
	      name = NULL;
	    }
	}

      if (args->hostname_dns_domain == 1)
        name = get_dns_domain_name (sname);
      else if (args->hostname_short == 1)
        name = get_short_hostname (sname);
    }
  else if (args->hostname_ip_address == 1)
      name = get_ip_addresses (sname);
  else
    {
      name = strdup (sname);
      if (name == NULL)
        error (EXIT_FAILURE, errno, "strdup");
    }

  puts (name);

  free (name);
  free (sname);
  return;
}

static void
set_name (const hostname_arguments *const args)
{
  char *hostname_new;
  int status;
  size_t size;

  if (args->hostname_file != NULL)
    hostname_new = parse_file (args->hostname_file);
  else
    hostname_new = args->hostname_new;

  size = strlen (hostname_new);
  if (!size)
    error (EXIT_FAILURE, 0, "Empty hostname");

  status = (*set_name_action) (hostname_new, size);
  if (status == -1)
    error (EXIT_FAILURE, errno, "sethostname");

  free (hostname_new);
  return;
}

static char *
get_aliases (const char *const host_name)
{
  char *aliases;
  unsigned int count = 0;
  unsigned int i;
  unsigned int size = 256;
  struct hostent *ht;

  aliases = xmalloc (sizeof (char) * size);
  aliases[0] = '\0';

  ht = gethostbyname (host_name);
  if (ht == NULL)
    strcpy (aliases, "(none)");
  else
    {
      for (i = 0; ht->h_aliases[i] != NULL; i++)
        {
          /* Aliases should be blankspace separated. */
          if (ht->h_aliases[i+1] != NULL)
            count++;
          count += strlen (ht->h_aliases[i]);
          if (count >= size)
            {
              size *= 2;
              aliases = xrealloc (aliases, size);
            }

          strcat (aliases, ht->h_aliases[i]);
          strcat (aliases, " ");
        }
    }

  return aliases;
}

static char *
get_fqdn (const char *const host_name)
{
  char *fqdn;
  struct hostent *ht;

  ht = gethostbyname (host_name);
  if (ht == NULL)
    fqdn = strdup ("(none)");
  else
    fqdn = strdup (ht->h_name);

  if (fqdn == NULL)
    error (EXIT_FAILURE, errno, "strdup");

  return fqdn;
}

static char *
get_ip_addresses (const char *const host_name)
{
  char address[16];
  char *addresses;
  unsigned int count = 0;
  unsigned int i;
  unsigned int size = 256;
  struct hostent *ht;

  addresses = xmalloc (sizeof (char) * size);
  addresses[0] = '\0';

  ht = gethostbyname (host_name);
  if (ht == NULL)
    strcpy (addresses, "(none)");
  else
    {
      for (i = 0; ht->h_addr_list[i] != NULL; i++)
        {
          inet_ntop (ht->h_addrtype, (void *) ht->h_addr_list[i],
                     address, sizeof (address));

          /* IP addresses should be blankspace separated. */
          if (ht->h_addr_list[i+1] != NULL)
            count++;
          count += strlen (address);
          if (count >= size)
            {
              size *= 2;
              addresses = xrealloc (addresses, size);
            }

          strcat (addresses, address);
          strcat (addresses, " ");
        }
    }

  return addresses;
}

static char *
get_dns_domain_name (const char *const host_name)
{
  char *domain_name;
  const char * pos;

  pos = strchr (host_name, '.');
  if (pos == NULL)
    domain_name = strdup ("(none)");
  else
    domain_name = strdup (pos+1);

  if (domain_name == NULL)
    error (EXIT_FAILURE, errno, "strdup");

  return domain_name;
}

static char *
get_short_hostname (const char *const host_name)
{
  size_t size;
  char *short_hostname;
  const char * pos;

  pos = strchr (host_name, '.');
  if (pos == NULL)
    short_hostname = strdup (host_name);
  else
    {
      size = pos - host_name;
      short_hostname = strndup (host_name, size);
    }

  if (short_hostname == NULL)
    error (EXIT_FAILURE, errno, "strdup");

  return short_hostname;
}

static char *
parse_file (const char *const file_name)
{
  char *buffer = NULL;
  char *name = NULL;
  FILE *file;
  ssize_t nread;
  size_t size = 0;

  file = fopen (file_name, "r");
  if (file == NULL)
    error (EXIT_FAILURE, errno, "fopen");

  errno = 0;			/* Portability issue!  */

  do
    {
      nread = getline (&buffer, &size, file);
      if (nread == -1)
	error (EXIT_FAILURE, errno, "getline%s", errno ? "" : ": No text");

      if (buffer[0] != '#')
        {
	  name = (char *) xmalloc (sizeof (char) * nread);
	  if (sscanf (buffer, "%s", name)  == 1)
	    break;
        }
    }
  while (feof (file) == 0);

  free (buffer);
  fclose (file);
  return name;
}
